### PR TITLE
Guidelines for writing unit tests

### PR DESCRIPTION
**Pull request by rjoleary**
Signed-off-by: Ryan O'Leary <ryan@ryanoleary.ca>

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -96,6 +96,98 @@ Every commit in your pull request needs to be able to build and pass the CI test
...
+func setMemAddr(addr, val uint, file string) error {
+	f, err := os.Open(file)
+	if err != nil {
+		return err
+	}
+	...
+}
+```
+
+2. Mocking objects:
```
Comment (rminnich) [inline CONTRIBUTING.md:149]: need to fix this as it is a dup of 'Mocking functions'

Comment (rminnich) [reply CONTRIBUTING.md:149]: This is an oldie but goodie, maybe worth referring to for file system mocks
https://talks.golang.org/2012/10things.slide#8

Comment (rjoleary) [reply CONTRIBUTING.md:149]: Done.

Comment (rjoleary) [reply CONTRIBUTING.md:149]: The new way to mock out filesystem is https://pkg.go.dev/io/fs and https://pkg.go.dev/testing/fstest. I put a mention above.

**Comment by rjoleary** [comment]
If this gets enough acceptance, I'll remove the "RFC:" from the commit message before submission.

**Review by rminnich — DISMISSED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -96,6 +96,120 @@ Every commit in your pull request needs to be able to build and pass the CI test
...
+### Mocking Dependencies
+
+* Special mocking packages should not be used.
+* Prefer to test the real thing where possible.
+* Mocking is sometimes necessary. For example:
+   * Operations as root.
+   * Interacting with special hardware (ex: USB, SPI, PCIe)
+   * Modifying machine state (ex: reboot, kexec)
+   * Tests which would otherwise be flaky (ex: `time.Sleep`, networking)
+* Consider writing an integration test for the above.
```
Comment (rminnich) [inline CONTRIBUTING.md:125]: This needs some explaining. 
* integration tests let you run the command under qemu, which lets you test operations with, e.g., virtual hardware. Consider writing an integration test if the program can not be easily run directly.

**Comment by 10000TB** [comment]
Note: Add a suggestion to put test data file created in test into tmp, `cmds/core/dd/dd_test.go` can fail in internal build/test run system, as it writes `datafile` to **current working directory**.

```
err := ioutil.WriteFile("datafile", []byte("ABCDEFG"), 0644)
```

https://github.com/u-root/u-root/blob/master/cmds/core/dd/dd_test.go#L235

**Comment by rjoleary** [comment]
> Note: Add a suggestion to put test data file created in test into tmp, `cmds/core/dd/dd_test.go` can fail in internal build/test run system, as it writes `datafile` to **current working directory**.
> 
> ```
> err := ioutil.WriteFile("datafile", []byte("ABCDEFG"), 0644)
> ```
> 
> https://github.com/u-root/u-root/blob/master/cmds/core/dd/dd_test.go#L235
> 
> sent a fix: #2201

Done!

File: CONTRIBUTING.md (hunk 2 above)
Comment (rjoleary) [reply CONTRIBUTING.md]: Fixed!

**Review by 10000TB — APPROVED** [review]

**Pull request merged**